From 15577a5cecf719c7afb0404f0095545d9c02a3be Mon Sep 17 00:00:00 2001
From: Siva Varalakshmi Devi <sivavaralakshmidevi@ami.com>
Date: Thu, 12 Dec 2024 15:21:06 +0530
Subject: [PATCH 15/15] OT-1596 Reconnect Support for RMedia Redirection
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In the event of a network issues on RMedia server, reconnect
will be initiated  based on the configured retry parameters.
Reconnection will be successfully established if the RMedia server’s
network is restored within the defined retry interval.

Signed-off-by: Siva Varalakshmi Devi <sivavaralakshmidevi@ami.com>
---
 meson.build                    |   1 +
 src/credentials.hpp            |  70 +++++++
 src/main.cpp                   |   3 +
 src/nfs.hpp                    |   6 +-
 src/resources.hpp              |   4 +-
 src/smb.hpp                    |   2 +-
 src/state/activating_state.cpp |  22 ++-
 src/system.hpp                 | 335 ++++++++++++++++++++++++++++++++-
 src/vm_interface.cpp           | 147 +++++++++++++++
 src/vm_interface.hpp           |  46 +++++
 virtual-media.json             |   4 +-
 11 files changed, 631 insertions(+), 9 deletions(-)
 create mode 100755 src/credentials.hpp
 create mode 100755 src/vm_interface.cpp
 create mode 100644 src/vm_interface.hpp

diff --git a/meson.build b/meson.build
index 6d8aeab..cdbfb37 100644
--- a/meson.build
+++ b/meson.build
@@ -206,6 +206,7 @@ configure_file(input : 'virtual-media.json',
 srcfiles_app = [ 'src/main.cpp',
                  'src/resources.cpp',
                  'src/state/activating_state.cpp',
+                 'src/vm_interface.cpp',
                ]
 
 bindir = get_option('prefix') + '/' +get_option('libexecdir')
diff --git a/src/credentials.hpp b/src/credentials.hpp
new file mode 100755
index 0000000..efe7aac
--- /dev/null
+++ b/src/credentials.hpp
@@ -0,0 +1,70 @@
+#pragma once
+
+//#include "logging.hpp"
+
+#include <boost/asio/buffer.hpp>
+#include <boost/asio/connect_pipe.hpp>
+#include <boost/asio/io_context.hpp>
+#include <boost/asio/readable_pipe.hpp>
+#include <boost/asio/writable_pipe.hpp>
+#include <boost/asio/write.hpp>
+
+#include <array>
+#include <string>
+
+// Wrapper for boost::async_pipe ensuring proper pipe cleanup
+class Credentials
+{
+  public:
+    explicit Credentials(boost::asio::io_context& io, std::string url = "", bool rw_status = false) : impl(io), read(io), url(std::move(url)), rw_status(rw_status)
+    {
+        boost::system::error_code ec;
+        boost::asio::connect_pipe(read, impl, ec);
+        if (ec)
+        {
+           // BMCWEB_LOG_CRITICAL("Failed to connect pipe {}", ec.what());
+        }
+    }
+
+    Credentials(const Credentials&) = delete;
+    Credentials(Credentials&&) = delete;
+    Credentials& operator=(const Credentials&) = delete;
+    Credentials& operator=(Credentials&&) = delete;
+
+    ~Credentials()
+    {
+        explicit_bzero(user.data(), user.capacity());
+        explicit_bzero(pass.data(), pass.capacity());
+    }
+
+    int releaseFd()
+    {
+        return read.release();
+    }
+
+    template <typename WriteHandler>
+    void asyncWrite(std::string&& username, std::string&& password,
+                    WriteHandler&& handler)
+    {
+        user = std::move(username);
+        pass = std::move(password);
+
+        // Add +1 to ensure that the null terminator is included.
+        std::array<boost::asio::const_buffer, 2> buffer{
+            {{user.data(), user.size() + 1}, {pass.data(), pass.size() + 1}}};
+        boost::asio::async_write(impl, buffer,
+                                 std::forward<WriteHandler>(handler));
+    }
+
+    std::string getUrl() const { return url; }
+    bool getRwStatus() const { return rw_status; }
+
+    boost::asio::writable_pipe impl;
+    boost::asio::readable_pipe read;
+
+  private:
+    std::string url;
+    bool rw_status;
+    std::string user;
+    std::string pass;
+};
diff --git a/src/main.cpp b/src/main.cpp
index ff13505..bceea9b 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -2,6 +2,7 @@
 #include "logger.hpp"
 #include "state_machine.hpp"
 #include "system.hpp"
+#include "vm_interface.hpp"
 
 #include <sys/mount.h>
 #include <sys/stat.h>
@@ -63,6 +64,8 @@ class App
     void run()
     {
         auto sessionMatch = dbusMonitor.sessionMonitor(bus);
+        vm::Interface interface(objServer);
+        interface.addInterfaces();
         ioc.run();
     }
 
diff --git a/src/nfs.hpp b/src/nfs.hpp
index 5c4d3b7..76a7e2d 100644
--- a/src/nfs.hpp
+++ b/src/nfs.hpp
@@ -17,7 +17,7 @@ class NfsShare
 
     bool mount(const fs::path& remote, bool rw)
     {
-        std::string options = "";
+        std::string options = "soft";
         std::string remoteNfsPath = "";
         std::string fs = "nfs4";
         unsigned long flags = (rw ? 0 : MS_RDONLY);
@@ -61,6 +61,10 @@ class NfsShare
         if (index != std::string::npos)
         {
             remoteNfsPath = remote.substr(index, remote.length() - 1);
+            if (!options.empty())
+            {
+                options += ",";
+            }
             options += ("addr=" + remote.substr(0, index));
             return true;
         }
diff --git a/src/resources.hpp b/src/resources.hpp
index 12d0a78..1b792d5 100644
--- a/src/resources.hpp
+++ b/src/resources.hpp
@@ -88,7 +88,7 @@ class Mount
 
     ~Mount()
     {
-        if (int result = ::umount(directory->getPath().string().c_str()))
+        if (int result = ::umount2(directory->getPath().string().c_str(), MNT_DETACH | MNT_FORCE ))
         {
             LogMsg(Logger::Error, result, " : Unable to unmout directory ",
                    directory->getPath());
@@ -128,7 +128,7 @@ class NfsMount
 
     ~NfsMount()
     {
-        if (int result = ::umount(directory->getPath().string().c_str()))
+        if (int result = ::umount2(directory->getPath().string().c_str(), MNT_DETACH | MNT_FORCE ))
         {
             LogMsg(Logger::Error, result, " : Unable to unmout directory ",
                    directory->getPath());
diff --git a/src/smb.hpp b/src/smb.hpp
index 05cd96a..ac499ad 100644
--- a/src/smb.hpp
+++ b/src/smb.hpp
@@ -21,7 +21,7 @@ class SmbShare
     {
         LogMsg(Logger::Debug, "Trying to mount remote : ", remote);
 
-        std::string options = "sec=ntlmsspi,seal,cache=none";
+        std::string options = "sec=ntlmsspi,seal,soft,cache=none,retrans=1,echo_interval=1";
         std::string credentialsOpt = "";
         const std::string fs = "cifs";
         std::string versionOpt = ",vers=3.1.1";
diff --git a/src/state/activating_state.cpp b/src/state/activating_state.cpp
index 37da28f..2f32b05 100644
--- a/src/state/activating_state.cpp
+++ b/src/state/activating_state.cpp
@@ -18,6 +18,9 @@
 #include <nlohmann/json.hpp>
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
+#include "credentials.hpp"
+boost::asio::io_context ioContext[2];
+std::shared_ptr<Credentials> creds[2];
 
 ActivatingState::ActivatingState(interfaces::MountPointStateMachine& machine) :
     BasicStateT(machine)
@@ -115,9 +118,23 @@ std::unique_ptr<BasicState> ActivatingState::activateLegacyMode()
                 "Failed to set parent permissions directory for socket");
         }
     }
-
+    
+    int slotNumber = extractSlotNumber(std::string(machine.getName()));
     if (isCifsUrl(machine.getTarget()->imgUrl))
     {
+        std::string user = machine.getTarget()->credentials->user();
+        std::string pass = machine.getTarget()->credentials->password();
+            creds[slotNumber] = std::make_shared<Credentials>(ioContext[slotNumber],machine.getTarget()->imgUrl, machine.getTarget()->rw);
+            creds[slotNumber]->asyncWrite(
+            std::move(user), std::move(pass),
+            [](const boost::system::error_code& ec,
+                                    std::size_t) {
+            if (ec)
+            {
+                LogMsg(Logger::Error, 
+                   " Unable to write the credentials");
+            }
+        }); 
         return mountSmbShare();
     }
     else if (isHttpsUrl(machine.getTarget()->imgUrl))
@@ -126,7 +143,8 @@ std::unique_ptr<BasicState> ActivatingState::activateLegacyMode()
     }
     else if (isNfsUrl(machine.getTarget()->imgUrl))
     {
-       return mountNfsShare();
+        creds[slotNumber] = std::make_shared<Credentials>(ioContext[slotNumber], machine.getTarget()->imgUrl, machine.getTarget()->rw);
+        return mountNfsShare();
     }
     return std::make_unique<ReadyState>(machine, std::errc::invalid_argument,
                                         "URL not recognized");
diff --git a/src/system.hpp b/src/system.hpp
index 8d7230f..a9c4648 100644
--- a/src/system.hpp
+++ b/src/system.hpp
@@ -17,8 +17,51 @@
 #include <sdbusplus/bus/match.hpp>
 #include <string>
 #include <variant>
-
+#include <sys/statfs.h>
+#include <thread>
+#include <chrono>
+#include <atomic>
+#include <cstdlib>
+#include <sdbusplus/bus.hpp>
+#include <sdbusplus/exception.hpp>
+#include <variant>
+static std::map<std::string, std::atomic<bool>> retryThreadCreatedMap;
+static std::atomic<bool> is_reconnecting;
 namespace fs = std::filesystem;
+#include "credentials.hpp"
+extern std::shared_ptr<Credentials> creds[2]; 
+using DbusVariantType = std::variant<
+    std::tuple<bool,std::string>,
+    std::vector<std::tuple<std::string, bool, bool, std::string>>,
+    std::tuple<bool, uint8_t, std::string>,	
+    std::vector<std::tuple<std::string, std::string, std::string>>,
+    std::vector<std::string>,
+    std::vector<double>,
+    std::string,
+    int64_t,
+    uint64_t,
+    double,
+    int32_t,
+    uint32_t,
+    int16_t,
+    uint16_t,
+    uint8_t,
+    bool,
+    sdbusplus::message::unix_fd,
+    std::vector<uint32_t>,
+    std::vector<uint16_t>,
+    sdbusplus::message::object_path,
+    std::tuple<uint64_t, std::vector<std::tuple<std::string, double, uint64_t>>>,
+    std::vector<sdbusplus::message::object_path>,
+    std::vector<std::tuple<std::string, std::string>>,
+    std::vector<std::tuple<uint32_t, bool, std::string>>,
+    std::vector<std::tuple<uint32_t, std::vector<uint32_t>>>,
+    std::vector<std::tuple<uint32_t, size_t>>,
+    std::vector<std::tuple<
+      std::vector<std::tuple<sdbusplus::message::object_path, std::string>>,
+      std::string, std::string, uint64_t>>
+ >;
+// Create an instance of the variant type
 
 #define POWER_SAVE_MODE_ENABLE 1
 #define POWER_SAVE_MODE_DISABLE 0
@@ -748,6 +791,279 @@ static int eject_status(const std::string& filePath)
     file.close();
     return slotNumber;
 }
+inline std::string getPathWithoutFileName(const std::string& path) {
+    size_t pos = path.find_last_of("/\\"); // Find last occurrence of '/' or '\'
+    if (pos != std::string::npos) {
+        return path.substr(0, pos + 1); // Return substring up to and including the last '/'
+    }
+    return ""; // If no delimiter found, return empty string
+}
+
+inline int isSamePath()
+{
+    if((creds[0] != NULL) && (creds[1] != NULL)){
+        std::string url_slot2  = getPathWithoutFileName(creds[0]->getUrl());
+        std::string url_slot3  = getPathWithoutFileName(creds[1]->getUrl());
+
+        if (url_slot2 == url_slot3) {
+            return 1;
+        } else {
+            return -1;
+        }
+    }
+    else
+    {
+        return 0;
+    }
+    
+}
+inline int isMountedPathAccessible(const std::string& path, int timeoutSeconds) {
+    std::atomic<int> isStatfsDone(0); // Unique atomic flag for each thread
+
+    std::thread statfsThread([&isStatfsDone, path]() {
+        struct statfs sb;
+        int ret = statfs(path.c_str(), &sb);
+        if (ret == -1) {
+            isStatfsDone = -1; // Path not accessible
+        } else {
+            isStatfsDone = 1;  // Accessible
+        }
+    });
+
+    for (int i = 0; i < timeoutSeconds; ++i) {
+        if (isStatfsDone != 0) {  // Check if statfs is done
+            statfsThread.join();
+            return isStatfsDone; // Return accessible state
+        }
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
+
+    if (isStatfsDone == 0) {
+        std::cerr << "Error: statfs() call timed out! for Path " << path << std::endl;
+        statfsThread.detach();
+        return false;
+    }
+
+    return true;
+}
+static bool isFileAccessible(const std::string& path) {
+    return fs::exists(path); // Check if the file exists
+}
+static bool getActiveStatus(const std::string& objectPath) {
+    auto bus = sdbusplus::bus::new_system();
+    try {
+        // Construct the method call message
+        sdbusplus::message::message reply = bus.new_method_call(
+            "xyz.openbmc_project.VirtualMedia",
+            objectPath.c_str(),
+            "org.freedesktop.DBus.Properties",
+            "Get");
+        reply.append("xyz.openbmc_project.VirtualMedia.Process", "Active");
+
+        // Prepare a variant to read the Active status
+        std::variant<bool> activeStatus;
+        bus.call(reply).read(activeStatus);
+        
+        // Return the boolean value from the variant
+        return std::get<bool>(activeStatus);
+    } catch (const sdbusplus::exception::SdBusError& e) {
+        LogMsg(Logger::Error, "Error retrieving VirtualMedia property: ", e.what());
+        return false;
+    }
+}
+static int extractSlotNumber(const std::string& path) {
+    // Find the position of the last underscore
+    size_t underscorePos = path.find_last_of('_');
+    int slotNumber = -1;
+    // Ensure underscore exists in the path
+    if (underscorePos != std::string::npos) {
+        // Extract the slot number after the underscore
+        slotNumber = std::stoi(path.substr(underscorePos + 1));
+        
+        // Normalize the slot number to be either 0 or 1
+        slotNumber = slotNumber % 2;
+    }
+    // Return the slot number
+    return slotNumber;
+}
+
+static bool retryMount(const std::string& localMountPath, unsigned int maxRetries, unsigned int retryDelay) {
+    // Variables to store mount configuration
+    std::string objpath = "/xyz/openbmc_project/VirtualMedia/Legacy/";
+    std::string interface = "xyz.openbmc_project.VirtualMedia.Legacy";
+    std::string slot_name = localMountPath;
+    DbusVariantType unixFd = -1;
+    std::string Local_image_name;
+    int slotNumber = extractSlotNumber(localMountPath);
+    int fd = -1;
+
+    if(creds[slotNumber] != NULL)
+    {
+        Local_image_name = "/tmp/" + localMountPath + "/" + creds[slotNumber]->getUrl().substr(creds[slotNumber]->getUrl().find_last_of("/\\") + 1);
+        objpath = objpath + localMountPath;
+
+        for (unsigned int attempt = 1; attempt <= maxRetries; ++attempt)
+        {
+            try {
+                LogMsg(Logger::Info, " Current Retry for ", localMountPath, " ... ", attempt);
+                if (getActiveStatus(objpath)) {
+                    LogMsg(Logger::Info, "Reconnect success for ", localMountPath); // To Make sure path is not accessible before going for the retry.
+                    return false;
+                }
+                
+                // Retry interval 
+		        std::this_thread::sleep_for(std::chrono::seconds(retryDelay));
+
+                // Reinitialize the bus and message for each attempt
+                int ret = -1;
+                auto b = sdbusplus::bus::new_system();
+                auto m = b.new_method_call(
+                    "xyz.openbmc_project.VirtualMedia",
+                    objpath.c_str(), interface.c_str(), "Mount");
+                std::string url = creds[slotNumber]->getUrl();
+                bool rwStatus = creds[slotNumber]->getRwStatus();
+
+                fd = creds[slotNumber]->releaseFd();
+                if (creds[slotNumber]->getUrl().find("nfs://") != 0) {
+                    unixFd = DbusVariantType(
+                        std::in_place_type<sdbusplus::message::unix_fd>, fd);
+                }
+                m.append(url);
+                m.append(rwStatus);
+                m.append(unixFd);
+                
+		        // Make the D-Bus call and read the status
+                auto reply = b.call(m);
+
+                //Check D-Bus status
+                for (unsigned int i = 0; i < 10; ++i) 
+                {
+                    if (getActiveStatus(objpath) || isFileAccessible(Local_image_name)) {
+                        LogMsg(Logger::Info, "Reconnect success for ", localMountPath);
+                        return false;
+                    }
+                    std::this_thread::sleep_for(std::chrono::seconds(1)); // Wait 1 second
+                }
+            }
+            catch (const sdbusplus::exception::SdBusError& e) {
+                LogMsg(Logger::Error, "D-Bus call failed with error: ", e.what());
+            } catch (const std::exception& e) {
+                LogMsg(Logger::Error, "Standard exception: ", e.what());
+            } catch (...) {
+                LogMsg(Logger::Error, "Unknown exception occurred during D-Bus call.");   
+            }
+        }
+    }
+    LogMsg(Logger::Info, "Error: Exhausted all retry attempts and failed to mounting on ", localMountPath);
+    creds[slotNumber] = nullptr;
+    return false;  // Failed to mount after all retries 
+}
+inline void handleUnmountAndRetry(const std::string& path, int slotNumber) 
+{
+    auto bus = sdbusplus::bus::new_system();
+    auto method = bus.new_method_call(
+        "xyz.openbmc_project.VirtualMedia",           // Service name
+        "/xyz/openbmc_project/VirtualMedia",          // Object path
+        "xyz.openbmc_project.VirtualMedia.Reconnect", // Interface
+        "GetAll"                                      // Method name
+    );
+
+    // Synchronously call the D-Bus method
+    auto reply = bus.call(method);
+
+    // Unpack the response directly
+    std::tuple<uint32_t, uint32_t> result;
+    reply.read(result);
+    unsigned int retryCount = std::get<0>(result);
+    unsigned int retryInterval = std::get<1>(result);
+
+    int ret = isSamePath();
+    if(ret == 1){
+        for (int i = 2; i <= 3; ++i) 
+        {   // Loop to unmount Slot_2 and Slot_3 if the path RMedia server is same
+            std::string slot = "Slot_" + std::to_string(i);
+            unMount(slot);
+            sleep(10);
+        }
+
+        for (int i = 2; i <= 3; ++i) 
+        {   // Loop for Rmedia reconnect for Slot_2 and Slot_3
+            std::string slot = "Slot_" + std::to_string(i);
+            if (!retryMount(slot, retryCount, retryInterval)) 
+            {
+                LogMsg(Logger::Debug, "Retry completed for ", path);
+            }
+            retryThreadCreatedMap[slot].store(false);
+        }
+    }
+    else{
+        unMount(path);
+        sleep(5);
+        if (!retryMount(path, retryCount, retryInterval)) 
+        {
+            LogMsg(Logger::Debug, "Retry completed for ", path);
+        }
+        retryThreadCreatedMap[path].store(false);
+    }
+}
+
+inline void mountMonitorThread(std::string path) {
+    int slotNumber = extractSlotNumber(path);
+    std::string mountPath = "/tmp/" + path;
+    
+    while (true) 
+    {
+	    int result = isMountedPathAccessible(mountPath, 5);
+        if(result == 0)
+        {
+            if (!retryThreadCreatedMap[path].load()) {
+
+            // Launch handleUnmountAndRetry in a separate detached thread
+                std::thread retryThread(handleUnmountAndRetry, path, slotNumber);
+                retryThread.detach();  // Detach to allow mountMonitorThread to continue without waiting
+                if(isSamePath() == 1)
+                {
+                    retryThreadCreatedMap["Slot_2"].store(true);
+                    retryThreadCreatedMap["Slot_3"].store(true);
+                }
+                else{
+                    retryThreadCreatedMap[path].store(true);
+                }
+                is_reconnecting.store(true);
+            }
+            break;  // Exit the loop to stop monitoring
+        } 
+
+        if ((result == -1) && (!retryThreadCreatedMap[path].load())) 
+        {
+            if(isSamePath() == 1){ // If both Slots uses same RMedia server then only one monitor thread will be created based on which Slot comes first. 
+            //If that slot redirection is stopped, Need to update SlotNumber and path to the other Slot. So, that it will continue to monitor.
+                if(slotNumber == 0){
+                    path = "Slot_3";
+                    mountPath = "/tmp/Slot_3";
+                    slotNumber = extractSlotNumber(path);
+                }
+                else if(slotNumber == 1){
+                    mountPath = "/tmp/Slot_2";
+                    path = "Slot_2";
+                    slotNumber = extractSlotNumber(path);
+                }
+                if (creds[slotNumber] != nullptr){
+                    creds[slotNumber] = nullptr;
+                }
+                continue;
+            }
+            else{
+                if (creds[slotNumber] != nullptr)
+                {
+                    creds[slotNumber] = nullptr;  
+                }  
+                break;
+            }
+        }
+        std::this_thread::sleep_for(std::chrono::seconds(5));
+	}
+}
 
 static void sleep_ms(int milliseconds)
 {
@@ -824,7 +1140,7 @@ struct UsbGadget
         uint8_t userId;
         bool status = false;
         int reason;
-        std::string mountingMethod;
+        std::string mountingMethod, mountpath;
 
         if (change == StateChange::inserted)
         {
@@ -877,6 +1193,21 @@ struct UsbGadget
                 auto reply = bus.call(msgReg);
                 reply.read(status);
 
+                mountpath = name;
+		        int slotNumber = extractSlotNumber(mountpath);
+
+                if(mountpath == "Slot_2" || mountpath == "Slot_3")
+                {
+                    if(creds[slotNumber] != NULL){
+                        if (creds[slotNumber]->getUrl().find("https://") != 0){ //To create monitor thread for CIFS/NFS
+                            if(( isSamePath() != 1) || is_reconnecting.load()){
+                                is_reconnecting.store(false);
+                                std::thread monitorThread(mountMonitorThread,mountpath);
+                                monitorThread.detach();
+                            }
+                        }
+                    }
+                }
                 if (status)
                 {
                     /* Get and update the SessionID in activeSessons */
diff --git a/src/vm_interface.cpp b/src/vm_interface.cpp
new file mode 100755
index 0000000..5035eef
--- /dev/null
+++ b/src/vm_interface.cpp
@@ -0,0 +1,147 @@
+#include <iostream>
+#include <fstream>
+#include "logger.hpp"
+#include "vm_interface.hpp"
+#include <nlohmann/json.hpp>
+
+#define MIN_RETRY_COUNT 3
+#define MAX_RETRY_COUNT 6
+#define MIN_RETRY_INTERVAL 15
+#define MAX_RETRY_INTERVAL 30
+
+using json = nlohmann::json;
+
+#define VIRTUAL_MEDIA_CONFIG_PATH "/etc/virtual-media.json"
+
+namespace vm
+{
+    // Declare a global or class-level json object to store the JSON data
+    json jsonData = {}; // Initialize with an empty JSON object
+
+    int Interface :: loadJson() {
+        try 
+        {
+            std::ifstream f(VIRTUAL_MEDIA_CONFIG_PATH);
+            if (!f.is_open()) {
+                throw std::runtime_error("Error opening JSON file");
+            }
+
+            jsonData = json::parse(f);
+
+            if (jsonData.is_null()) {
+                throw std::runtime_error("JSON data is empty");
+            }
+        } catch (const std::exception &e) {
+            LogMsg(Logger::Error, e.what());
+            return -1;
+        }
+        return 0;
+    }
+
+    // Constructor to initialize the object server and add the interface
+    Interface::Interface(std::shared_ptr<sdbusplus::asio::object_server> objServer) : server(objServer)
+    {}
+
+    // Method to add interfaces
+    void Interface::addInterfaces()
+    {
+        loadJson();
+        addRmediaInterface();
+    }
+
+    // Method to add the Virtual Media interface
+    void Interface::addRmediaInterface()
+    {
+        vmediaInterface = server->add_interface(vmObjPath.c_str(), RmediaInterface.c_str());
+
+        RetryCount    = jsonData["RetryCount"];
+        RetryInterval = jsonData["RetryInterval"];
+
+        // Register the property for RetryCount
+        vmediaInterface->register_property(
+            "RetryCount", RetryCount,
+            sdbusplus::asio::PropertyPermission::readOnly);
+
+        // Register the property for RetryInterval
+        vmediaInterface->register_property(
+            "RetryInterval", RetryInterval,
+            sdbusplus::asio::PropertyPermission::readOnly);
+
+        // Register the method to set both RetryCount and RetryInterval
+        vmediaInterface->register_method("SetAll", [this](unsigned int RetryCount, unsigned int RetryInterval) {
+           return SetAll(RetryCount, RetryInterval);
+        });
+
+        // Register the method to get both RetryCount and RetryInterval
+        vmediaInterface->register_method("GetAll", [this]() {
+            return GetAll();
+        });
+
+        // Initialize the interface to finalize its setup
+        vmediaInterface->initialize();
+    }
+
+    // Method to set the retry parameters with validation
+    std::string Interface::SetAll(unsigned int newRetryCount, unsigned int newRetryInterval)
+    {
+        std::string status = "Unknown";
+        
+        // Validate the new values for retry count and interval
+        if ((newRetryCount < MIN_RETRY_COUNT || newRetryCount > MAX_RETRY_COUNT) || (newRetryInterval < MIN_RETRY_INTERVAL || newRetryInterval > MAX_RETRY_INTERVAL))
+        {
+            status = "Error: RetryCount must be set between 3 and 6. RetryInterval should be configured between 15 and 30.";
+            return status;
+        }
+
+        // Check if there is any change in the values
+        if (RetryCount == newRetryCount && RetryInterval == newRetryInterval)
+        {
+            status = "Success";
+            return status;
+        }
+
+        // Update the member variables
+        RetryCount = newRetryCount;
+        RetryInterval = newRetryInterval;
+
+        // Update D-Bus properties if the interface is available
+        if (vmediaInterface)
+        {
+            vmediaInterface->set_property("RetryCount", RetryCount);
+            vmediaInterface->set_property("RetryInterval", RetryInterval);
+            status = "Success";
+        }
+
+        // Load JSON data only if needed
+        try
+        {
+           loadJson();  // Load JSON only once
+
+            // Update JSON only if both values differ
+            if (jsonData["RetryCount"] != RetryCount || jsonData["RetryInterval"] != RetryInterval)
+            {
+                jsonData["RetryCount"] = RetryCount;
+                jsonData["RetryInterval"] = RetryInterval;
+                std::ofstream outputFile(VIRTUAL_MEDIA_CONFIG_PATH);
+                if (outputFile.is_open()) {
+                    outputFile << jsonData.dump(4);  // Write updated JSON to file
+                    outputFile.close();
+                } else {
+                    LogMsg(Logger::Error, "Error in writing the the file! ", VIRTUAL_MEDIA_CONFIG_PATH);
+                }
+            }
+        }
+        catch (const std::exception& e)
+        {
+            LogMsg(Logger::Error, "Error loading JSON file:!", e.what());
+        }
+        return status;
+    }
+
+    // Method to get the retry parameters
+    std::tuple<unsigned int, unsigned int> Interface::GetAll() const
+    {
+        return std::make_tuple(RetryCount, RetryInterval);
+    }
+
+} // namespace vm
diff --git a/src/vm_interface.hpp b/src/vm_interface.hpp
new file mode 100644
index 0000000..998e97a
--- /dev/null
+++ b/src/vm_interface.hpp
@@ -0,0 +1,46 @@
+#ifndef INTERFACE_HPP
+#define INTERFACE_HPP
+
+#include <sdbusplus/asio/object_server.hpp>
+#include <memory>
+#include <string>
+#include <tuple>
+
+namespace vm
+{
+    const std::string vmObjPath = "/xyz/openbmc_project/VirtualMedia";
+    const std::string RmediaInterface = "xyz.openbmc_project.VirtualMedia.Reconnect";
+
+    class Interface
+    {
+    public:
+        // Constructor to initialize the object server and add the interface
+        Interface(std::shared_ptr<sdbusplus::asio::object_server> objServer);
+
+        // Method to load JSON configuration
+        static int loadJson();  // Add this line to declare the loadJson method
+
+        // Method to add interfaces
+        void addInterfaces();
+
+    private:
+        // Method to add the Virtual Media interface
+        void addRmediaInterface();
+
+        // Method to set the retry parameters with validation
+        std::string SetAll(unsigned int RetryCount, unsigned int RetryInterval);
+
+        // Method to get the retry parameters
+        std::tuple<unsigned int, unsigned int> GetAll() const;
+
+        // Member variables
+        std::shared_ptr<sdbusplus::asio::object_server> server;
+        std::shared_ptr<sdbusplus::asio::dbus_interface> vmediaInterface; // D-Bus interface object
+
+        unsigned int RetryCount = 0;     // Initial default value for retry count
+        unsigned int RetryInterval = 0;  // Initial default value for retry interval
+    };
+
+} // namespace vm
+
+#endif // INTERFACE_HPP
diff --git a/virtual-media.json b/virtual-media.json
index 688b74a..ca66c42 100644
--- a/virtual-media.json
+++ b/virtual-media.json
@@ -33,5 +33,7 @@
             "Timeout": 90,
             "BlockSize": 512
         }
-    }
+    },
+    "RetryCount": 3,
+    "RetryInterval": 15
 }
-- 
2.34.1

